File: app/lib/services/expense_service.dart
import '../database/database_helper.dart';
import '../models/expense.dart';

class ExpenseService {
  final DatabaseHelper _db = DatabaseHelper();

  ({int start, int end}) _monthRange(DateTime date) {
    final start = DateTime(date.year, date.month, 1).millisecondsSinceEpoch;
    final end = DateTime(date.year, date.month + 1, 1).millisecondsSinceEpoch;
    return (start: start, end: end);
  }

  Future<List<Expense>> getExpensesForMonth([DateTime? date]) async {
    final d = date ?? DateTime.now();
    final range = _monthRange(d);
    final db = await _db.database;
    final rows = await db.query(
      'expenses',
      where: 'created_at >= ? AND created_at < ?',
      whereArgs: [range.start, range.end],
      orderBy: 'created_at DESC',
    );
    return rows.map(Expense.fromMap).toList();
  }

  Future<double> getExpensesTotalForMonth([DateTime? date]) async {
    final list = await getExpensesForMonth(date);
    return list.fold<double>(0.0, (sum, e) => sum + e.amount);
  }

  Future<double> getTrueProfit(double grossProfit, [DateTime? date]) async {
    final expenses = await getExpensesTotalForMonth(date);
    return grossProfit - expenses;
  }

  Future<Expense> addExpense({
    required String category,
    required double amount,
    String? note,
    int? createdAt,
  }) async {
    final db = await _db.database;
    final expense = Expense(
      category: category,
      amount: amount,
      note: note?.trim().isEmpty == true ? null : note?.trim(),
      createdAt: createdAt ?? DateTime.now().millisecondsSinceEpoch,
    );
    final id = await db.insert('expenses', expense.toMap());
    return Expense(
      id: id,
      category: expense.category,
      amount: expense.amount,
      note: expense.note,
      createdAt: expense.createdAt,
    );
  }

  Future<void> deleteExpense(int id) async {
    final db = await _db.database;
    await db.delete('expenses', where: 'id = ?', whereArgs: [id]);
  }
}

File: web/src/lib/expenseService.ts
import { db, monthRangeFromDate } from './db';
import type { Expense, ExpenseCategory } from '@/types';

export async function getExpensesForMonth(date = new Date()): Promise<Expense[]> {
  const { start, end } = monthRangeFromDate(date);
  return db.expenses
    .where('created_at')
    .between(start, end, true, false)
    .reverse()
    .sortBy('created_at');
}

export async function getExpensesTotalForMonth(date = new Date()): Promise<number> {
  const list = await getExpensesForMonth(date);
  return list.reduce((sum, e) => sum + e.amount, 0);
}

export async function getTrueProfit(grossProfit: number, date = new Date()): Promise<number> {
  const expenses = await getExpensesTotalForMonth(date);
  return grossProfit - expenses;
}

export async function addExpense(input: {
  category: ExpenseCategory;
  amount: number;
  note?: string;
  created_at?: number;
}): Promise<Expense> {
  const expense: Expense = {
    category: input.category,
    amount: input.amount,
    note: input.note?.trim() || null,
    created_at: input.created_at ?? Date.now(),
  };
  const id = await db.expenses.add(expense);
  return { ...expense, id };
}

export async function deleteExpense(id: number): Promise<void> {
  await db.expenses.delete(id);
}
